light-client/relayer/src/relayer.rs

```diff
@@ -0,0 +1,286 @@
+    /// Entrypoint for relayer loop.
+    pub fn start(
+        config: Config,
+        ledger_db: LedgerDB,
+        watcher_db: WatcherDB,
+        sender: S,
+        verifier: V,
+        stop_requested: Arc<AtomicBool>,
+        shared_state: Arc<Mutex<RelayerSharedState>>,
+        logger: Logger,
+    ) {
+        let next_block_index = config.start_block_index;
+        let thread = Self {
+            _config: config,
+            next_block_index,
+            ledger_db,
+            watcher_db,
+            sender,
+            verifier,
+            stop_requested,
+            shared_state,
+            logger,
+        };
+        thread.run();
+    }
+
+    pub fn run(mut self) {
+        // Poll ledger for data to relay
+        log::info!(self.logger, "Relayer thread started.");
+        loop {
+            self.shared_state().current_block_index = self.next_block_index;
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Relayer thread stop requested.");
+                break;
+            }
+            match self.ledger_db.get_block_data(self.next_block_index) {
+                Err(LedgerError::NotFound) => std::thread::sleep(Self::POLLING_FREQUENCY),
+                Err(e) => {
+                    log::error!(
+                        self.logger,
+                        "Unexpected error when checking for block data {}: {:?}",
+                        self.next_block_index,
+                        e
+                    );
+                    std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                }
+                Ok(block_data) => {
+                    if let Err(err) = self.process_block(&block_data) {
+                        log::error!(
+                            self.logger,
+                            "When processing block {}: {:?}",
+                            self.next_block_index,
+                            err
+                        );
+                        std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    } else {
+                        self.next_block_index += 1;
+                    }
+                }
+            }
+        }
+    }
+
+    fn process_block(&mut self, block_data: &BlockData) -> Result<(), Error> {
+        let tracer = tracer!();
+
+        let mut span =
+            block_span_builder(&tracer, "poll_block", self.next_block_index).start(&tracer);
+
+        span.set_attribute(TELEMETRY_BLOCK_INDEX_KEY.i64(self.next_block_index as i64));
+
+        let relevant_burns = Self::check_for_relevant_burns(&block_data.contents().outputs);
+        if relevant_burns.is_empty() {
+            return Ok(());
+        }
+
+        // Try to get signatures from the watcher
+        tracer.in_span("loop_for_signatures", |_cx| loop {
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Relayer thread stop requested.");
+                return Ok(());
```

Comment:
This is a bit suboptimal with regards to updating `next_block_index` outside this function.
Right now its going to be okay, but down the road we're going to want the relayer to store the last processed block into a file so that when the process restarts it doesn't re-scan the entire blockchain (similar to `start_from` in [ledger-distribution](https://github.com/mobilecoinfoundation/mobilecoin/blob/master/ledger/distribution/src/main.rs)).
The easy solution might be to move `self.next_block_index += 1;` inside this method